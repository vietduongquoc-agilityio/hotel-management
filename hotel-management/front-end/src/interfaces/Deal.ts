export interface DealData {
  dealName: string;
  referenceNumber: string;
  startDate: Date;
  endDate: Date;
  roomType: string;
  statusDeal: string;
  reservationsLeft: number;
  documentId: string;
}

export type NewDealData = Omit<DealData, "documentId">;
