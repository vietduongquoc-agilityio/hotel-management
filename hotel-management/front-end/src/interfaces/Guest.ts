export interface GuestData {
    guestName: string;
    roomType: string;
    stay: number;
    price: number;
    registrationNumber: string;
    totalAmount: number;
    checkInDate: Date;
    documentId: string;
}

export type NewGuestData = Omit<GuestData, "documentId">;
