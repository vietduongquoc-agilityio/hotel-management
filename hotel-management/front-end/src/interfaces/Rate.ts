export interface RateData {
  roomType: string;
  cancellationPolicy: string;
  deals: string;
  dealPrice: string;
  availability: string;
  documentId: string;
  rate: string;
  totalOfBooked?: number;
}

export type NewRateData = Omit<RateData, "documentId">;
