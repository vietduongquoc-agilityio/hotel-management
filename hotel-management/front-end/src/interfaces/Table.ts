export interface EditFunctionType<T> {
  initialData: T;
  onEditFunction: (updatedData: T) => void;
}

export interface DeleteFunctionType {
  documentId: string;
  onDeleteFunction: (id: string) => void;
}
