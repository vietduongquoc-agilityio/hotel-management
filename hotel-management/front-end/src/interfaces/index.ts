export * from "./Rate";
export * from "./Room";
export * from "./Table";
export * from "./Guest"
