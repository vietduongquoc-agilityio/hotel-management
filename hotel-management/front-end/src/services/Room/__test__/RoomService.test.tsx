import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import {
  getRooms,
  createRoomApi,
  getRoomById,
  updateRoom,
  deleteRoom,
} from "@/services";
import { NewRoomData } from "@/interfaces";

const mock = new MockAdapter(axios);
const BASE_URL = process.env.VITE_BASE_URL;

describe("roomService", () => {
  afterEach(() => {
    mock.reset();
  });

  describe("getRooms", () => {
    it("should fetch rooms with pagination data", async () => {
      const mockResponse = {
        data: [
          { id: "1", attributes: { name: "Room 1" } },
          { id: "2", attributes: { name: "Room 2" } },
        ],
        meta: { pagination: { page: 1, pageSize: 10, pageCount: 1, total: 2 } },
      };

      mock.onGet(`${BASE_URL}/rooms`).reply(200, mockResponse);

      const result = await getRooms(1, 10);
      expect(result).toMatchSnapshot();
    });

    it("should show error toast on failure", async () => {
      mock.onGet(`${BASE_URL}/rooms`).reply(500);

      await expect(getRooms(1, 10)).rejects.toThrow(
        "Request failed with status code 500"
      );
    });
  });

  describe("createRoomApi", () => {
    it("should create a new room", async () => {
      const roomData: NewRoomData = {
        bedType: "Room 1",
        roomStatus: "Available",
        roomFacility: "",
        roomFloor: "",
        roomNumber: "",
      };
      const mockResponse = { data: { id: "1", attributes: { ...roomData } } };

      mock.onPost(`${BASE_URL}/rooms`).reply(200, mockResponse);

      const result = await createRoomApi(roomData);
      expect(result).toEqual(mockResponse);
    });

    it("should show error toast on failure", async () => {
      const roomData: NewRoomData = {
        bedType: "Room 1",
        roomStatus: "Available",
        roomFacility: "",
        roomFloor: "",
        roomNumber: "",
      };
      mock.onPost(`${BASE_URL}/rooms`).reply(500);

      await expect(createRoomApi(roomData)).rejects.toThrow(
        "Request failed with status code 500"
      );
    });
  });

  describe("getRoomById", () => {
    it("should fetch room by ID", async () => {
      const roomId = "1";
      const mockResponse = {
        data: { id: roomId, attributes: { name: "Room 1" } },
      };

      mock.onGet(`${BASE_URL}/rooms/${roomId}`).reply(200, mockResponse);

      const result = await getRoomById(roomId);
      expect(result).toEqual(mockResponse);
    });

    it("should throw error for invalid room ID", async () => {
      const roomId = "1";
      mock.onGet(`${BASE_URL}/rooms/${roomId}`).reply(404);

      await expect(getRoomById(roomId)).rejects.toThrow(
        "Request failed with status code 404"
      );
    });
  });

  describe("updateRoom", () => {
    it("should update a room by ID", async () => {
      const roomId = "1";
      const roomData: NewRoomData = {
        bedType: "Updated Room",
        roomStatus: "Booked",
        roomFacility: "",
        roomFloor: "",
        roomNumber: "",
      };
      const mockResponse = {
        data: { id: roomId, attributes: { ...roomData } },
      };

      mock.onPut(`${BASE_URL}/rooms/${roomId}`).reply(200, mockResponse);

      const result = await updateRoom(roomId, roomData);
      expect(result).toEqual(mockResponse);
    });
  });

  describe("deleteRoom", () => {
    it("should delete a room by ID", async () => {
      const roomId = "1";
      const mockResponse = { message: "Room deleted successfully" };

      mock.onDelete(`${BASE_URL}/rooms/${roomId}`).reply(200, mockResponse);

      const result = await deleteRoom(roomId);
      expect(result).toEqual(mockResponse);
    });

    it("should throw error for invalid room ID", async () => {
      const roomId = "1";
      mock.onDelete(`${BASE_URL}/rooms/${roomId}`).reply(404);

      await expect(deleteRoom(roomId)).rejects.toThrow(
        "Request failed with status code 404"
      );
    });
  });
});
