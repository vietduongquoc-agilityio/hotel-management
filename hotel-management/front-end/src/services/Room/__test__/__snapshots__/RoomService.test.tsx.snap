// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roomService getRooms should fetch rooms with pagination data 1`] = `
{
  "pagination": {
    "page": 1,
    "pageCount": 1,
    "pageSize": 10,
    "total": 2,
  },
  "rooms": [
    {
      "attributes": {
        "name": "Room 1",
      },
      "id": "1",
    },
    {
      "attributes": {
        "name": "Room 2",
      },
      "id": "2",
    },
  ],
}
`;
